python/splines.cpp

```diff
@@ -0,0 +1,240 @@
+
+    // Bind the interpolation type enum that lets us set different interpolation types for our splines
+    py::enum_<SplineInterpolation>(m_splines, "InterpolationType")
+            .value("Linear", SplineInterpolation::kLinear, "Linear interpolation between the knots")
+            .value("Linear_Func", SplineInterpolation::kLinearFunc, "Same as 'Linear'")
+            .value("Cubic_TSpline3", SplineInterpolation::kTSpline3, "Use same coefficients as `ROOT's TSpline3 <https://root.cern.ch/doc/master/classTSpline3.html>`_ implementation")
+            .value("Cubic_Monotonic", SplineInterpolation::kMonotonic, "Coefficients are calculated such that the segments between knots are forced to be monotonic. The implementation we use is based on `this method <https://www.jstor.org/stable/2156610>`_ by Fritsch and Carlson.")
+            .value("Cubic_Akima", SplineInterpolation::kAkima, "The second derivative is not required to be continuous at the knots. This means that these splines are useful if the second derivative is rapidly varying. The implementation we used is based on `this paper <http://www.leg.ufpr.br/lib/exe/fetch.php/wiki:internas:biblioteca:akima.pdf>`_ by Akima.")
+            .value("N_Interpolation_Types", SplineInterpolation::kSplineInterpolations, "This is only to be used when iterating and is not a valid interpolation type.");
+
+    
+    py::class_<SplineBase, PySplineBase /* <--- trampoline*/>(m_splines, "SplineBase");
+
+    py::class_<TResponseFunction_red>(m_splines, "_ResponseFunctionBase")
+        .doc() = "Base class of the response function, this binding only exists for consistency with the inheritance structure of the c++ code. Just pretend it doesn't exist and don't worry about it...";
+
+    // Bind the TSpline3_red class. Decided to go with a clearer name of ResponseFunction for the python binding
+    // and make the interface a bit more python-y. Additionally remove passing root stuff so we don't need to deal 
+    // with root python binding and can just pass it native python objects.
+    py::class_<TSpline3_red, TResponseFunction_red, std::unique_ptr<TSpline3_red, py::nodelete>>(m_splines, "ResponseFunction")
+        .def(
+            // define a more python friendly constructor that massages the inputs and passes them
+            // through to the c++ constructor
+            py::init
+            (
+                // Just take in some vectors, then build a TSpline3 and pass this to the constructor
+                [](std::vector<double> &xVals, std::vector<double> &yVals, SplineInterpolation interpType)
+                {
+                    if ( xVals.size() != yVals.size() )
+                    {
+                        throw MaCh3Exception(__FILE__, __LINE__, "Different number of x values and y values!");
+                    }
+
+                    int length = xVals.size();
+
+                    if (length == 1)
+                    {
+                        _float_ *pars[3];
```

Comment:
Is there a reason this isn't an std::vector?

Reply:
nope